lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,291 @@
+
+const {
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_HTTP_TOKEN,
+    ERR_INVALID_THIS,
+    ERR_HTTP_INVALID_HEADER_VALUE
+  }
+} = require('internal/errors');
+
+const {
+  ArrayIsArray,
+  ArrayPrototypeSlice,
+  ArrayPrototypeSplice,
+  MathFloor,
+  ObjectEntries,
+  RegExpPrototypeSymbolReplace,
+  StringPrototypeLocaleCompare,
+  StringPrototypeToLocaleLowerCase,
+  Symbol,
+  SymbolIterator,
+  ObjectDefineProperties
+} = primordials;
+
+const { validateObject } = require('internal/validators');
+const { isBoxedPrimitive } = require('internal/util/types');
+
+const { validateHeaderName, validateHeaderValue } = require('_http_outgoing');
+
+const { Buffer } = require('buffer');
+
+const { inspect } = require('internal/util/inspect');
+
+const kHeadersList = Symbol('headers list');
+
+/**
+ * This algorithm is based off of
+ * https://www.tbray.org/ongoing/When/200x/2003/03/22/Binary
+ * It only operates on the even indexes of the array (the header names) by only
+ * iterating at most half the length of the input array. The search also
+ * assumes all entries are strings and uses String.prototype.localeCompare for
+ * comparison.
+ */
+function binarySearch(arr, val) {
+  let low = 0;
+  let high = MathFloor(arr.length / 2);
+
+  while (high > low) {
+    const mid = (high + low) >>> 1;
+
+    if (StringPrototypeLocaleCompare(val, arr[mid * 2]) > 0) {
+      low = mid + 1;
+    } else {
+      high = mid;
+    }
+  }
+
+  return low * 2;
+}
+
+function normalizeAndValidateHeaderName(name) {
+  if (name === undefined) {
+    throw new ERR_INVALID_HTTP_TOKEN('Header name', name);
+  }
+  const normalizedHeaderName = StringPrototypeToLocaleLowerCase(name);
+  validateHeaderName(normalizedHeaderName);
+  return normalizedHeaderName;
+}
+
+function normalizeAndValidateHeaderValue(name, value) {
+  if (value === undefined) {
+    throw new ERR_HTTP_INVALID_HEADER_VALUE(value, name);
+  }
+  // https://fetch.spec.whatwg.org/#concept-header-value-normalize
+  const normalizedHeaderValue = RegExpPrototypeSymbolReplace(
+    /^[\n\t\r\x20]+|[\n\t\r\x20]+$/g,
+    value,
+    ''
+  );
+  validateHeaderValue(name, normalizedHeaderValue);
+  return normalizedHeaderValue;
+}
+
+function isHeaders(object) {
+  return kHeadersList in object;
+}
+
+function fill(headers, object) {
+  if (isHeaders(object)) {
+    // Object is instance of Headers
+    headers[kHeadersList] = ArrayPrototypeSlice(object[kHeadersList]);
+  } else if (ArrayIsArray(object)) {
+    // Support both 1D and 2D arrays of header entries
+    if (ArrayIsArray(object[0])) {
+      // Array of arrays
+      for (let i = 0; i < object.length; i++) {
+        if (object[i].length !== 2) {
+          throw new ERR_INVALID_ARG_VALUE('init', object[i],
+                                          'is not of length 2');
+        }
+        headers.append(object[i][0], object[i][1]);
+      }
+    } else if (typeof object[0] === 'string' || Buffer.isBuffer(object[0])) {
```

Comment:
Browsers don't seem to support this:

```js
new Headers(['test', 'value'])
// Chromium: Uncaught TypeError: Failed to construct 'Headers': The provided value cannot be converted to a sequence.
// Firefox: Uncaught TypeError: Headers constructor: Element of sequence<sequence<ByteString>> branch of (sequence<sequence<ByteString>> or record<ByteString, ByteString>) can't be converted to a sequence.
// Safari: TypeError: Value is not a sequence
```
Am I missing something or should this be removed? Also why `Buffer`?

Reply:
This was the one major deviation from spec. Supporting a 1D array of headers proves to be the most efficient solution

Reply:
This might be something that is worth bringing back to the WHATWG spec folks to see if 1D arrays *could* be supported. /cc @annevk @domenic

Reply:
It would be best if someone filed an issue against whatwg/fetch to discuss that, ideally with motivation. It seems a bit too special to me, but it might fly.

Reply:
I think the idea is to match JS's Map type, so probably first you'd want to get Ecma 262's Map updated.

Reply:
I'd also suggest not adding non-spec constructors if you want this capability but instead some Node-specific utility method e.g. `http.headersFrom1DArray()`. The way you've done it now promotes non-isomorphic code.

Reply:
I'll open something in the fetch repo. I'm unclear how this parallels to ECMA Map though.

Reply:
The 2D array syntax for the Headers constructor is derived from the 2D array syntax for the Map constructor.

Reply:
going to resolve this comment and we can continue the discussion in the whatwg/fetch repo